py/builtinevex.c

```diff
@@ -114,12 +114,18 @@ STATIC mp_obj_t eval_exec_helper(size_t n_args, const mp_obj_t *args, mp_parse_i
     mp_obj_dict_t *globals = mp_globals_get();
     mp_obj_dict_t *locals = mp_locals_get();
     if (n_args > 1) {
-        globals = MP_OBJ_TO_PTR(args[1]);
-        if (n_args > 2) {
+        if (args[1] != mp_const_none) {
+            globals = MP_OBJ_TO_PTR(args[1]);
+        }
+        if (n_args > 2 && args[2] != mp_const_none) {
             locals = MP_OBJ_TO_PTR(args[2]);
         } else {
             locals = globals;
         }
+        if (!MP_OBJ_IS_TYPE(globals, &mp_type_dict) ||
+            !MP_OBJ_IS_TYPE(locals, &mp_type_dict)) {
```

Comment:
The unix nanbox build is failing because the first arg to MP_OBJ_IS_TYPE must be of type mp_obj_t, not a pointer.  Some refactoring of the code will fix it, probably without any change in code size.